src/Statement/StatementListChanger.php

```diff
@@ -0,0 +1,169 @@
+<?php
+
+namespace Wikibase\DataModel\Services\Statement;
+
+use Wikibase\DataModel\Entity\PropertyId;
+use Wikibase\DataModel\Statement\Statement;
+use Wikibase\DataModel\Statement\StatementList;
+
+/**
+ * A collection of higher level utility functions to manipulate StatementList objects.
+ *
+ * @since 3.7
+ *
+ * @license GPL-2.0+
+ * @author Thiemo Mättig
+ */
+class StatementListChanger {
+
+	/**
+	 * @var StatementList
+	 */
+	private $statementList;
+
+	public function __construct( StatementList $statementList ) {
+		$this->statementList = $statementList;
+	}
+
+	/**
+	 * Removes all statements from the $statementList object given in the constructor.
+	 */
+	public function clear() {
+		foreach ( $this->statementList->toArray() as $statement ) {
+			$this->statementList->removeStatementsWithGuid( $statement->getGuid() );
+		}
+	}
+
+	/**
+	 * Makes sure all statements with the same property are next to each other (forming a group),
+	 * and reorders them if necessary. The position of the group in the list is determined by the
+	 * first statement with the same property.
+	 */
+	public function groupByProperty() {
+		$byId = [];
+
+		foreach ( $this->statementList->toArray() as $statement ) {
+			$id = $statement->getPropertyId()->getSerialization();
+			$byId[$id][] = $statement;
+		}
+
+		$this->clear();
+
+		foreach ( $byId as $statements ) {
+			foreach ( $statements as $statement ) {
+				$this->statementList->addStatement( $statement );
+			}
+		}
+	}
+
+	/**
+	 * @param Statement $newStatement
+	 * @param int|null $index An absolute index in the list. If the index is not next to a statement
+	 *  with the same property, the closest possible position is used instead. Default is null,
+	 *  which adds the new statement after the last statement with the same property, or to the end.
+	 */
+	public function addToGroup( Statement $newStatement, $index = null ) {
+		$statements = $this->statementList->toArray();
+		$id = $newStatement->getPropertyId();
+
+		if ( $index === null ) {
+			$index = $this->getLastIndexWithinGroup( $statements, $id );
+		} else {
+			// Limit search range to avoid looping non-existing positions
+			$validIndex = min( max( 0, $index ), count( $statements ) );
+			$index = $this->getClosestIndexWithinGroup( $statements, $id, $validIndex );
+			if ( $index === null ) {
+				$index = $this->getClosestIndexAtGroupBorder( $statements, $validIndex );
+			}
+		}
```

Comment:
Not sure what you mean with "crap". Can you please avoid using such offensive language?

@Benestar suggested to replace our naive StatementList with one that's aware of PropertyId groups a long time ago, see #507. I'm not opposing such a change.

I still believe the code proposed here is a helpful step in a longer migration path that might end with replacing StatementList.

Note that the code I'm proposing here does have a very specific advantage: It allows us to add new statements where they belong, without the need to reorder existing statement lists. Not that this is of much relevance, because users usually do not see diffs on the JSON level. But still. This smooth migration is not possible with a StatementList that enforces grouping on construction time.